### ISSUE TITLE
Question: how to listen for final states in React components?

### ISSUE THREAD

**Issue by jbreckmckye** [davidkpiano/xstate#1031]
This is a somewhat open-ended question and is arguably a bit of an X/Y problem.

Today I was trying to write a UI component that used XState to model its state.

The component was a kind of date picker. It could return either a Date or a null representing whether or when an event had occurred. The UI had two states: a low-resolution state (has the event occurred?) and a high-resolution state (when did it occur?)

I wanted my machine to have a `DONE` state representing the user choosing an option, submitting it and it being validated. That way my state machine documented the 'whole flow', so to speak.

To give a rough impression...

```typescript
// keeping this as simple as I can
export default createMachine<Ctx, Events, States>({
  initial: 'LOW_RES',
  states: {
    LOW_RES: {
      on: {
        ANSWER_OCCURRED: 'HIGH_RES',
        ANSWER_SUBMIT: 'DONE' // skipping guards, validations etc. to keep the code example short
      }
    },
    HIGH_RES: {
      on: {
        ANSWER_NOT_OCCURRED: 'LOW_RES',
        ANSWER_SUBMIT: 'DONE'
      }
    },
    DONE: {
      type: 'final',
      data: ctx => ctx
    }
  }
})
```

But in my React component, how do I invoke a callback when the `DONE` state was reached?

I tried using `onDone`...

```typescript
// again, I am simplifying a lot here
const [state, send, service] = useMachine(fsm)

service.onDone(() => {
  props.submit(state.context) // or similar
})
```

But I had a problem! `onDone` seems to fire multiple times! This caused me some trouble, as I didn't really want `props.submit()` to be invoked more than once. And from what I've read, the parent state of a statemachine isn't supposed to fire done events anyway.

And yet, the typescript bindings' docs suggest otherwise - that onDone is for the whole statechart. Is this true? I'm a bit unclear about what SCXML actually mandates here.

![image](https://user-images.githubusercontent.com/3148617/75372051-c2e4e500-58bf-11ea-8f1f-a9eb7ee37688.png)

There is another approach I found, where I could pass `actions` into useMachine...

```
useMachine(fsm, {
  actions: {
    onDone: (ctx) => onSubmit(ctx)
  }
})
```

But this seems a little bit bulky compared to `onDone`.

**tl;dr**

How should I solve this problem, in the main? Am I mis-using state machines? Is 'decorating' actions, as per the second approach, the XState idiom?

And outside of React - how does one start a state machine and then subscribe to its completion?

Any advice is appreciated - I'm very keen to use XState in my next project.

**Comment by davidkpiano** [comment]
This is the problem:

```js

// set right in the function scope
service.onDone(() => {
  props.submit(state.context) // or similar
})

```

Because you're calling `service.onDone(...)` in the component function scope, it will be called _on every rerender_, which means you will be registering that `props.submit` function more than once.

There's two solutions to this:

1. Idiomatic React: make this call in `useEffect()`:

```js

useEffect(() => {
  const listener = () => {
    props.submit(state.context) // or similar
  };

  service.onDone(listener);

  return () => { service.off(listener) }
}, [service, props.submit]);

```

2. Use an action in the machine instead, in your `'DONE'` state:

```js
// in the machine
DONE: {
  type: 'final',
  entry: 'submit'
}
```
```js
// in the component
const [state, send] = useMachine(machine, {
  actions: {
    submit(ctx) {
      props.submit(ctx);
    }
  }
});
```

**Comment by jbreckmckye** [comment]
Ah! I'm an idiot! I forgot to wrap my call in a `useEffect`! So therefore every re-render will re-bind the function call. That explains it.

Thank you for answering so swiftly.
